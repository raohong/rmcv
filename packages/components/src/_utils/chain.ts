const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
