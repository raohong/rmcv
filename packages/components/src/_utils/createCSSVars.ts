import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
