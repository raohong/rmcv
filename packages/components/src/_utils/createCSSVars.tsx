import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
