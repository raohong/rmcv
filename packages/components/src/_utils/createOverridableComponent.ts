import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
