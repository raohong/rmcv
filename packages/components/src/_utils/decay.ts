import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
