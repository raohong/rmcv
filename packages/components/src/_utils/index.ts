export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
