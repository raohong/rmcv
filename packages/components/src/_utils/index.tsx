export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
