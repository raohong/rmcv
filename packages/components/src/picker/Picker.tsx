'use-client';

import { useControllableValue, useUpdateEffect } from '@rmc-vant/hooks';
import { compose, isFunction, omit } from '@rmc-vant/utils';
import clsx from 'clsx';
import React, { useMemo } from 'react';
import { elementOpacityHapticFeedback } from '../_styles';
import { useThemeProps } from '../config-provider';
import PickerColumn from './PickerColumn';
import { PickerName, composePickerSlotClassNames } from './classNames';
import type {
  PickerBaseOptionWithChildren,
  PickerComponentState,
  PickerProps,
  PickerValue,
} from './interface';
import {
  PickerCancelButton,
  PickerColumnList,
  PickerConfirmButton,
  PickerIndicator,
  PickerLoading,
  PickerLoadingRoot,
  PickerMask,
  PickerPopup,
  PickerRoot,
  PickerTitle,
  PickerToolbar,
} from './styles';
import {
  fillValueByNestedOptions,
  getPickerColumnData,
  getPickerColumnsLength,
  isNestedOptions,
} from './utils';

const feedback = elementOpacityHapticFeedback();

const Picker = <V extends PickerValue>(
  props: PickerProps<V>,
  ref: React.Ref<HTMLDivElement>,
) => {
  const {
    title,
    cancelButtonText,
    confirmButtonText,
    onCancel,
    onColumnChange,
    defaultValue,
    onChange,
    className,
    lazyRender,
    onOpen,
    children,
    classNames,
    loading,
    immediateChange = true,
    columns = [],
    showToolbar = true,
    visibleOptionNum = 6,
    optionHeight = 44,
    toolbarPosition = 'top',
    popup = false,
    ...rest
  } = useThemeProps(PickerName, props);

  const [value, setValue] = useControllableValue(rest, {
    defaultValue,
  });
  const [open, setOpen] = useControllableValue(rest, {
    valuePropName: 'open',
    trigger: 'onOpenChange',
    defaultValue: false,
  });

  const length = useMemo(() => getPickerColumnsLength(columns), [columns]);
  const isNested = isNestedOptions(columns);
  const { columns: internalColumns, indexList } = useMemo(
    () => getPickerColumnData(value, columns, length),
    [columns, value, length],
  );

  const componentState: PickerComponentState = useMemo(
    () => ({
      popup,
      toolbarPosition,
      loading: !!loading,
    }),
    [popup, toolbarPosition, loading],
  );
  const slotClassNames = composePickerSlotClassNames(componentState, classNames);

  const handleChange = (columnIndex: number, val: V) => {
    onColumnChange?.(columnIndex, val);

    const next = value ? [...value] : [];
    const previousLength = next.length;
    next[columnIndex] = val;

    if (previousLength > length) {
      for (let i = previousLength - 1; i >= length; i--) {
        next.splice(i, 1);
      }
    }

    for (let i = 0; i < next.length; i++) {
      if (isNested && i > columnIndex) {
        // @ts-ignore
        next[i] = undefined;
      }
      else if (i !== columnIndex && next[i] === undefined) {
        next[i] = internalColumns[i][0].value;
      }
    }

    // 嵌套结构置 0
    if (isNested) {
      fillValueByNestedOptions(
        next,
        columns as PickerBaseOptionWithChildren<V>[],
        length,
      );
    }

    onChange?.(next, columnIndex);

    setValue(next);
  };

  useUpdateEffect(() => {
    if (!popup) {
      setOpen(false);
    }
  }, [popup]);

  const handleClose = () => {
    setOpen(false);
  };

  const handleCancel = () => {
    if (popup) {
      handleClose();
    }

    onCancel?.();
  };

  const handleConfirm = () => {
    if (popup) {
      handleClose();
    }

    rest?.onConfirm?.(value);
  };

  const handleOpen = () => {
    setOpen(true);
    onOpen?.();
  };

  const totalHeight = visibleOptionNum * optionHeight;
  const content = (
    <PickerRoot
      className={clsx(slotClassNames.root, className)}
      {...omit(rest, ['value', 'onConfirm', 'children'])}
      ref={ref}
      componentState={componentState}
    >
      {!!showToolbar && (
        <PickerToolbar
          componentState={componentState}
          className={slotClassNames.toolbar}
        >
          <PickerCancelButton
            componentState={componentState}
            className={slotClassNames.cancelButton}
            activeStyle={feedback}
            onClick={handleCancel}
          >
            {cancelButtonText ?? '取消'}
          </PickerCancelButton>
          <PickerTitle
            className={slotClassNames.title}
            componentState={componentState}
          >
            {title}
          </PickerTitle>
          <PickerConfirmButton
            componentState={componentState}
            className={slotClassNames.confirmButton}
            activeStyle={feedback}
            onClick={handleConfirm}
          >
            {confirmButtonText ?? '确定'}
          </PickerConfirmButton>
        </PickerToolbar>
      )}
      {internalColumns.length > 0 && (
        <PickerColumnList>
          {internalColumns.map((list, index) => (
            <PickerColumn<V>
              // eslint-disable-next-line react/no-array-index-key
              key={index}
              selectedIndex={indexList[index]}
              options={list}
              optionHeight={optionHeight}
              onChange={handleChange}
              totalHeight={totalHeight}
              columnIndex={index}
              immediateChange={immediateChange}
              componentState={componentState}
              slotClassNames={slotClassNames}
            />
          ))}
          <PickerMask
            style={{
              backgroundSize: `100% ${totalHeight / 2 - optionHeight / 2}px`,
            }}
            className={slotClassNames.mask}
            componentState={componentState}
          />
          <PickerIndicator
            style={{ height: optionHeight }}
            className={slotClassNames.indicator}
            componentState={componentState}
          />
        </PickerColumnList>
      )}
      {loading && (
        <PickerLoadingRoot>
          <PickerLoading
            className={slotClassNames.loading}
            componentState={componentState}
          />
        </PickerLoadingRoot>
      )}
    </PickerRoot>
  );

  const child
    = popup && children
      ? isFunction(children) ? children(value) : children
      : null;

  return popup
    ? (
        <>
          {!!child
          && React.cloneElement(child!, {
            onClick: compose(handleOpen, child?.props.onClick),
          })}
          <PickerPopup
            className={slotClassNames.popup}
            position='bottom'
            open={open}
            onClose={handleClose}
            lazyRender={lazyRender}
            componentState={componentState}
            round
          >
            {content}
          </PickerPopup>
        </>
      )
    : (
        <>{content}</>
      );
};

export default React.forwardRef(Picker) as <T extends PickerValue>(
  props: PickerProps<T>,
) => JSX.Element;
