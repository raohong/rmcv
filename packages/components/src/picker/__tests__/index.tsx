import { render, screen } from '@testing-library/react';
import React, { useState } from 'react';
import userEvent from '@testing-library/user-event';
import { getPrefixCls } from '../../_utils';
import Picker from '..';

const options = ['A', 'B', 'C'].map((item) => ({ value: item }));

const testId = 'picker';

test('render correctly', () => {
  const tree = render(<Picker columns={[['A']]} />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with title', () => {
  render(<Picker title="test-title" data-testid={testId} />);

  expect(screen.getByText('test-title')).toBeInTheDocument();
});

test('render with showToolbar', () => {
  render(<Picker showToolbar={false} data-testid={testId} />);

  expect(
    screen.getByTestId(testId).querySelector(`.${getPrefixCls('picker-toolbar')}`),
  ).not.toBeInTheDocument();
});

test('render with confirmButtonText', () => {
  render(<Picker confirmButtonText="ok" data-testid={testId} />);

  expect(screen.getByText('ok')).toBeInTheDocument();
});

test('render with cancelButtonText', () => {
  render(<Picker cancelButtonText="cancel" data-testid={testId} />);

  expect(screen.getByText('cancel')).toBeInTheDocument();
});

test('render with toolbarPosition', () => {
  render(<Picker toolbarPosition="bottom" data-testid={testId} />);

  expect(screen.getByTestId(testId)).toHaveClass(
    getPrefixCls('picker-position-bottom'),
  );
});

test('render with internalControllable', () => {
  const onConfirm = jest.fn();
  render(<Picker onConfirm={onConfirm} columns={[options]} data-testid={testId} />);

  const confirm = screen
    .getByTestId(testId)
    .querySelector(`.${getPrefixCls('picker-confirm-button')}`)!;

  userEvent.click(confirm);

  expect(onConfirm).toBeCalledWith(['A']);
});

describe('render with value', () => {
  it('internalControllable', () => {
    const onConfirm = jest.fn();

    render(
      <Picker columns={[options]} data-testid={testId} onConfirm={onConfirm} />,
    );

    const option = screen
      .getByTestId(testId)
      .querySelector(`.${getPrefixCls('picker-option')}:nth-child(2)`)!;

    userEvent.click(option);

    expect(option).toHaveAttribute('data-selected', 'true');

    const confirm = screen
      .getByTestId(testId)
      .querySelector(`.${getPrefixCls('picker-confirm-button')}`)!;

    userEvent.click(confirm);

    expect(onConfirm).toBeCalledWith(['B']);
  });
});
