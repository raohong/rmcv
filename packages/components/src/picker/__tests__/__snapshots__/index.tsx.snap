// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmcv-picker"
  >
    <div
      class="rmcv-picker-toolbar"
    >
      <button
        class="rmcv-picker-button rmcv-picker-cancel-button rmcv-touchable"
        style="cursor: cursor;"
      >
        取消
      </button>
      <h4
        class="rmcv-picker-title"
      />
      <button
        class="rmcv-picker-button rmcv-picker-confirm-button rmcv-touchable"
        style="cursor: cursor;"
      >
        确定
      </button>
    </div>
    <div
      class="rmcv-picker-column-container"
    >
      <div
        class="rmcv-picker-column"
        style="height: 264px;"
      >
        <div
          class="rmcv-picker-column-slider"
        >
          <ul
            class="rmcv-picker-column-slider-content"
            style="top: calc(50% - 22px); transform: none;"
          >
            <li
              class="rmcv-picker-option"
              data-selected="true"
              role="button"
              style="height: 44px;"
              tabindex="0"
            >
              A
            </li>
          </ul>
        </div>
      </div>
      <div
        class="rmcv-picker-mask"
        style="background-size: 100% 110px;"
      />
      <div
        class="rmcv-picker-indicator"
        style="height: 44px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
